clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -979,6 +1013,13 @@ void handleCoordinatorReady() {
                 null;
         this.coordinatorSupportsBumpingEpoch = initProducerIdVersion != null &&
                 initProducerIdVersion.maxVersion() >= 3;
+
+        // TODO(caliu) use feature version.
+        ApiVersion produceVersion = nodeApiVersions != null ?
+                nodeApiVersions.apiVersion(ApiKeys.PRODUCE) :
+                null;
+        this.coordinatorSupportsTransactionV2 = produceVersion != null &&
```

Comment:
I think the other thing we are looking for is in the txn offset commit request and the produce request, setting the version to 11 and 4 until v2/feature version is enabled. I didn't see that here yet.

Reply:
Good point! Will check.

Reply:
Sorry I think I used some unclear wording. When I said "checking TV", I meant using the nodeApiVersions to check the TV on requests -- this will be in ApiVersionsResponseData.finalizedFeatures. We will also want to check ApiVersionsResponseData.finalizedFeaturesEpoch when we update the feature version. I was thinking the transaction manager could have some sort of central mechanism that checks the api version and updates on the various requests. This could be done in a helper.